v2/spiffeid/path.go

```diff
@@ -0,0 +1,106 @@
+package spiffeid
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// ValidatePath validates that a path string is a conformant path for a SPIFFE
+// ID. Namely:
+// - does not contain an empty segments (including a trailing slash)
+// - does not contain dot segments (i.e. '.' or '..')
+// - does not contain any percent encoded characters
+// - has only characters from the unreserved or sub-delims set from RFC3986.
+func ValidatePath(path string) error {
+	if reason := validatePath(path); reason != noReason {
+		return pathErr{path: path, reason: reason}
+	}
+	return nil
+}
+
+func validatePath(path string) errReason {
+	switch {
+	case path == "":
+		return noReason
+	case path[0] != '/':
+		return noLeadingSlash
+	}
+
+	segmentStart := 0
+	segmentEnd := 0
+	for ; segmentEnd < len(path); segmentEnd++ {
+		c := path[segmentEnd]
+		if c == '/' {
+			switch path[segmentStart:segmentEnd] {
+			case "/":
+				return emptySegment
+			case "/.", "/..":
+				return dotSegment
+			}
+			segmentStart = segmentEnd
+			continue
+		}
+		if !isValidPathSegmentChar(c) {
+			return badPathSegmentChar
+		}
+	}
+
+	switch path[segmentStart:segmentEnd] {
+	case "/":
+		return trailingSlash
+	case "/.", "/..":
+		return dotSegment
+	}
+	return noReason
+}
+
+// isValidPathSegmentChar returns true if the given character is acceptable in
+func formatPath(format string, args ...interface{}) (string, error) {
+	path := fmt.Sprintf(format, args...)
+	if err := ValidatePath(path); err != nil {
+		return "", err
+	}
+	return path, nil
+}
+
+func joinPathSegments(segments ...string) (string, error) {
+	buf := new(bytes.Buffer)
```

Comment:
👍 habitual, i usually use `new` if I'm passing it into functions that take `io.Writer` to avoid having to write `&buf` everywhere (escape analysis will prevent it from going on the heap either way). But that isn't the case here. Happy to change.